REGRESSION (280975@main) Null pointer deref crash in WebCore::AudioWorkletGlobalScope::createProcessor

rreno opened the pull request:
#### 2ef9bf886adc838c1e70bec46da48a5e33c9b7e8
<pre>
REGRESSION (280975@main) Null pointer deref crash in WebCore::AudioWorkletGlobalScope::createProcessor
<a href="https://bugs.webkit.org/show_bug.cgi?id=278512">https://bugs.webkit.org/show_bug.cgi?id=278512</a>
<a href="https://rdar.apple.com/133250806">rdar://133250806</a>

Reviewed by Chris Dumez.

After 280975@main, Audio Worklet processor constructors are weak
handles and may be garbage collected prematurely, especially in WK1.
This can lead to null pointer deref crashes in WebAudio WPT when trying
to construct a new Audio Worklet processor.

This change adds the AudioWorkletGlobalScope as a WebCore opaque root in
order to keep registered processor constructors alive via the
isReachableFromOpaqueRoots mechanism.

* Source/WebCore/bindings/js/JSAudioWorkletGlobalScopeCustom.cpp:
(WebCore::JSAudioWorkletGlobalScope::visitAdditionalChildren):

Canonical link: <a href="https://commits.webkit.org/282644@main">https://commits.webkit.org/282644@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9fae741ee3490a99a4fc917439e7fdf78b98ee2d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/63713 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/43070 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/16310 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/67734 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/14321 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/65833 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/50757 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/14601 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/51329 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/9898 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/66782 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/39920 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/55150 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/31961 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/36601 "Found 1 new test failure: imported/w3c/web-platform-tests/websockets/basic-auth.any.sharedworker.html?wss (failure)") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/12529 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/13194 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/58430 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/12854 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/69430 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/7660 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/12421 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/58606 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/7693 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/55243 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/58818 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14114 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/6367 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/38890 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/39969 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/41081 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/39712 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/Modules/webaudio/AudioWorkletGlobalScope.cpp:
```diff
@@ -117,7 +120,11 @@ ExceptionOr<void> AudioWorkletGlobalScope::registerProcessor(String&& name, Ref<
+    decltype(m_processorConstructorMap)::AddResult addResult;
+    {
+        Locker locker { m_processorConstructorMapLock };
+        addResult = m_processorConstructorMap.add(name, WTFMove(processorConstructor));
```
I recommend we use `bool isNewEntry` variable instead of `addResult`. `addResult` contains an iterator which would be unsafe to use outside the scope of the lock, so it's a foot gun.

rreno replied:
Great point. Given your other comment though we can probably avoid all this locking if I get the worklet's context added as an opaque root instead.

cdumez commented on Source/WebCore/Modules/webaudio/AudioWorkletProcessorConstructor.idl:
```diff
@@ -28,5 +28,4 @@
 
 [
     Conditional=WEB_AUDIO,
-    GenerateIsReachable=ImplScriptExecutionContext
```
It seems this has no effect on callbacks, shouldn't we fix that instead? If I look at the generated code for JSAudioWorkletProcessorConstructor, it definitely doesn't generate the `isReachableFromOpaqueRoots` function I'd expect it to. I suspect the bindings generator ignores `GenerateIsReachable` for callbacks and only considers it for interfaces.

One of the reasons I'm concerned is:
- how about IdleRequestCallback, doesn't it have the same bug for example? It also uses `GenerateIsReachable=ImplScriptExecutionContext` and the generated bindings code also doesn't contain a isReachableFromOpaqueRoots function.

rreno replied:
The code for `GenerateIsReachable` for callbacks is a bit different from interfaces. For callbacks all it does is it constructs the `JSC::WeakImpl` with a pointer to the `ScriptExecutionContext` instead of a pointer to `this` (the callback). In the main thread the script execution context has been added as an opaque root to the visitor so it works there. Since its not working here and I went back to make sure the code generator was working like I meant it to it must be that nothing is adding the AudioWorklet's context as an opaque root to the visitor. So that is what probably needs to be fixed here.

rreno replied:
For completeness, the `isReachableFromOpaqueRoots` implementation is in `WebCore/bindings/js/JSCallbackData.h` and so the code generator doesn't need to generate the WeakHandleOwner stuff  like it does for other IDL types.

rreno replied:
Yeah so adding `addWebCoreOpaqueRoot(visitor, wrapped())` to `JSAudioWorkletGlobalScope::visitAdditionalChildren` fixes this with no need to lock the map. Since its not possible to unregister a processor constructor after registering (as far as I can tell) that one line change is all that's needed to fix the bug.

cdumez commented on Source/WebCore/bindings/js/JSAudioWorkletGlobalScopeCustom.cpp:
```diff
@@ -36,6 +37,8 @@ namespace WebCore {
 template<typename Visitor>
 void JSAudioWorkletGlobalScope::visitAdditionalChildren(Visitor& visitor)
 {
+    addWebCoreOpaqueRoot(visitor, wrapped());
```
For extra safety, I think we should do:
```
addWebCoreOpaqueRoot(visitor, static_cast<ScriptExecutionContext&>(wrapped()));
```
Since opaque roots work as `void*`, they lose all typing information and casting to the base class may be important in cases where the pointer value ends up being different (though likely not the case here).

This would also be consistent with what we do for Document:
```
template<typename Visitor>
void JSDocument::visitAdditionalChildren(Visitor& visitor)
{
    addWebCoreOpaqueRoot(visitor, static_cast<ScriptExecutionContext&>(wrapped()));
}
```

cdumez reviewed (approved):
LGTM with nit.

rreno committed:
REGRESSION (280975@main) Null pointer deref crash in WebCore::AudioWorkletGlobalScope::createProcessor
https://bugs.webkit.org/show_bug.cgi?id=278512
rdar://133250806

Reviewed by Chris Dumez.

After 280975@main, Audio Worklet processor constructors are weak
handles and may be garbage collected prematurely, especially in WK1.
This can lead to null pointer deref crashes in WebAudio WPT when trying
to construct a new Audio Worklet processor.

This change adds the AudioWorkletGlobalScope as a WebCore opaque root in
order to keep registered processor constructors alive via the
isReachableFromOpaqueRoots mechanism.

* Source/WebCore/bindings/js/JSAudioWorkletGlobalScopeCustom.cpp:
(WebCore::JSAudioWorkletGlobalScope::visitAdditionalChildren):

Canonical link: https://commits.webkit.org/282644@main
